### PR TITLE
feat: memory server

### PR DESCRIPTION
**Pull request by wendytang**
Brings forward memory server. 

**Example Usage**: 
- "remember the fitness goal for 2025 is to do x pushups" -> remember tool call
- "what was my fitness goal?" -> retrieve tool call
- "i don't want to do the pushup goal anymore, please erase it" -> remove specific memory

image

## Notes
- Memory server is implemented with tool calls. 
- Remembered files are added as Active Resources for when Resources are incorporated. This can also be easily removed.
- (in-progress)Testing in session

### CODE REVIEW

**Comment by Kvadratni** [comment]
I love this more refined approach to the same simple file based idea.

**Comment by salman1993** [comment]
I tested this out with `goosed` but it didn't work for me. I tried with both the `default` and the `reference` agent versions. This is what I tried:

1. checkout this branch and merge in latest changes - `git merge origin/v1.0` 
2. create an agent 
```
curl --request POST \
  --url http://localhost:3000/agent \
  --header 'Content-Type: application/json' \
  --header 'X-Secret-Key: test' \
  --data '{
  "version": "default",  # also tried "reference" here
  "provider": "databricks"
}'
```

3. add memory system
```
curl --request POST \
  --url http://localhost:3000/systems/add \
  --header 'Content-Type: application/json' \
  --header 'X-Secret-Key: test' \
  --data '{
  "type": "Stdio",
  "cmd": "cargo", 
  "args": ["run", "--bin", "goosed", "--", "mcp", "memory"]
}'
```

4. remember memory - this succeeds  
```
curl --request POST \
  --url http://localhost:3000/reply \
  --header 'Accept: text/event-stream' \
  --header 'Content-Type: application/json' \
  --header 'X-Secret-Key: test' \
  --header 'x-protocol: data' \
  --data '{
  "messages": [
    {
      "role": "user",
      "content": "remember my global goal of 20 pushups a day for category '\''fitness'\''"
    }
  ]
}'
```
Screenshot 2025-01-16 at 11 02 20 AM

5. retrieve memory - this fails with 2 possible errors - (1) data must be string, (2) server error
```
curl --request POST \
  --url http://localhost:3000/reply \
  --header 'Accept: text/event-stream' \
  --header 'Content-Type: application/json' \
  --header 'X-Secret-Key: test' \
  --header 'x-protocol: data' \
  --data '{
  "messages": [
    {
      "role": "user",
      "content": "retrieve my global memories for category '\''fitness'\''"
    }
  ]
}'
```

error - data must be a string:
Screenshot 2025-01-16 at 11 02 58 AM


error - Server error: Server error occured. Status: 503 Service Unavailable
Screenshot 2025-01-16 at 10 48 27 AM

also see these logs in goosed agent:
Screenshot 2025-01-16 at 10 49 55 AM

File: crates/goose-mcp/src/memory/mod.rs [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+        if category.is_empty() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Category must be a string",
+            ));
+        }
+
+        let data = args["data"]
+            .as_str()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Data must be a string"))?;
```
Comment (salman1993) [inline crates/goose-mcp/src/memory/mod.rs]: i think this fails for `retrieve_memories` which doesn't have a 'data' property in the JSON schema

**Comment by salman1993** [comment]
the fix was easy - needed to make data field optional. i put the changes here if you wanna merge to this branch - https://github.com/block/goose/pull/629

i'm not sure why the Inspector tool did not catch this. i think it'd be good to test this out more thoroughly via `goosed`, here are some things i'd recommend testing:
1. try adding 'developer2' and 'memory' system. ask questions that require using a tool from each of those systems
2. edge cases - try to make the memory fail with confusing categories or is_global or not
3. retrieving a memory 5-6 msgs later in the conversation (this will test out the prompting in some sense)

**Comment by wendytang** [comment]
tested memory on a few more examples, it seems to work as expected. I think generally if we want to improve the usage of memory tool, we could consider reducing friction in the tool use by removing category and global arguments.
